fpga_interchange/arch.h

```diff
@@ -758,39 +818,102 @@ struct Arch : ArchAPI<ArchRanges>
+            result.first->second.boundcells.resize(tile_type.bel_data.size());
+            result.first->second.tags.resize(default_tags.size());
+        }
+
+        return result.first->second;
+    }
+
     void bindBel(BelId bel, CellInfo *cell, PlaceStrength strength) override
     {
         NPNR_ASSERT(bel != BelId());
-        NPNR_ASSERT(tileStatus[bel.tile].boundcells[bel.index] == nullptr);
 
-        tileStatus[bel.tile].boundcells[bel.index] = cell;
+        TileStatus &tile_status = get_tile_status(bel.tile);
+        NPNR_ASSERT(tile_status.boundcells[bel.index] == nullptr);
+
+        const auto &bel_data = bel_info(chip_info, bel);
+        NPNR_ASSERT(bel_data.category == BEL_CATEGORY_LOGIC);
+
+        if(io_ports.count(cell->type) == 0) {
+            int32_t mapping = bel_info(chip_info, bel).pin_map[get_cell_type_index(cell->type)];
+            NPNR_ASSERT(mapping >= 0);
+
+            if (cell->cell_mapping != mapping) {
+                map_cell_pins(cell, mapping);
+            }
+            constraints.bindBel(tile_status.tags.data(), get_cell_constraints(bel, cell->type));
+        } else {
+            map_port_pins(bel, cell);
+            // FIXME: Probably need to actually constraint io port cell/bel,
+            // but the current BBA emission doesn't support that.  This only
+            // really matters if the placer can choose IO port locations.
+        }
+
+        tile_status.boundcells[bel.index] = cell;
+
         cell->bel = bel;
         cell->belStrength = strength;
+
+
         refreshUiBel(bel);
     }
 
     void unbindBel(BelId bel) override
     {
         NPNR_ASSERT(bel != BelId());
-        NPNR_ASSERT(tileStatus[bel.tile].boundcells[bel.index] != nullptr);
-        tileStatus[bel.tile].boundcells[bel.index]->bel = BelId();
-        tileStatus[bel.tile].boundcells[bel.index]->belStrength = STRENGTH_NONE;
-        tileStatus[bel.tile].boundcells[bel.index] = nullptr;
+
+        TileStatus &tile_status = get_tile_status(bel.tile);
+        NPNR_ASSERT(tile_status.boundcells[bel.index] != nullptr);
+
+        CellInfo *cell = tile_status.boundcells[bel.index];
+        tile_status.boundcells[bel.index] = nullptr;
+
+        cell->bel = BelId();
+        cell->belStrength = STRENGTH_NONE;
+
+        // FIXME: Probably need to actually constraint io port cell/bel,
+        // but the current BBA emission doesn't support that.  This only
+        // really matters if the placer can choose IO port locations.
+        if(io_ports.count(cell->type) == 0) {
+            constraints.unbindBel(tile_status.tags.data(), get_cell_constraints(bel, cell->type));
+        }
+
         refreshUiBel(bel);
     }
 
-    bool checkBelAvail(BelId bel) const override { return tileStatus[bel.tile].boundcells[bel.index] == nullptr; }
+    bool checkBelAvail(BelId bel) const override
+    {
+        // FIXME: This could consult the constraint system to see if this BEL
```

Comment:
this should probably wait until we've had a more general think about placement validity - use of `checkBelAvail` and `getConflictingBelCell` for validity reasons isn't really something we've done before but does have some interesting applications, though would need some placer work to take advantage of